### PR TITLE
[AutoDiff] Support differentiation of side effects.

### PR DESCRIPTION
**Pull request by rxwei**
This patch adds support for differentiating through reads and writes from/to `var`s. Specifically, what made this possible is a set of differentiation rules for side-effecting operations: `alloc_stack`, `dealloc_stack`, `begin_access`, `end_access`, `load`, `store`.

This patch also fixes the following minor bugs:

* When `AdjointEmitter::getAdjointValue` creates a default adjoint value, correctly remap the original type to get rid of archetypes.

* Handle more reference counting instructions.

* Emit better diagnostics for not being able to differentiate writes to global variables and to mutable captures.

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -2777,25 +2784,52 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
...
+  /// Materialize an adjoint value in the most efficient way.
+  SILValue materializeAdjoint(AdjointValue val, SILLocation loc);
+
   /// Given two adjoint values, accumulate them.
   AdjointValue accumulateAdjointsDirect(AdjointValue lhs, AdjointValue rhs);
 
   /// Given two materialized adjoint values, accumulate them. These two
   /// adjoints must be objects of loadable type.
   SILValue accumulateMaterializedAdjointsDirect(SILValue lhs, SILValue rhs);
+
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: extra newline

**Review by dan-zheng — APPROVED** [review]
👍

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
Tests passing locally. Merging to unblock deadline.

**Pull request merged**